man/man8/zstream.8
.\" SPDX-License-Identifier: CDDL-1.0
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or https://opensource.org/licenses/CDDL-1.0.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright (c) 2020 by Delphix. All rights reserved.
.\"
.Dd May 12, 2026
.Dt ZSTREAM 8
.Os
.
.Sh NAME
.Nm zstream
.Nd manipulate ZFS send streams
.Sh SYNOPSIS
.Nm
.Cm dump
.Op Fl Cvd
.Op Ar file
.Nm
.Cm decompress
.Op Fl v
.Op Ar object Ns Sy \&, Ns Ar offset Ns Op Sy \&, Ns Ar type Ns ...
.Nm
.Cm drop_record
.Op Fl v
.Op Ar object Ns Sy \&, Ns Ar offset Ns Op Sy \& Ns ...
.Nm
.Cm redup
.Op Fl v
.Ar file
.Nm
.Cm token
.Ar resume_token
.Nm
.Cm raw
.Op Fl v
.Op Fl g Ar fromguid
.Ar image|device
.Nm
.Cm recompress
.Op Fl t Ar num_threads
.Op Fl l Ar level
.Ar algorithm
.
.Sh DESCRIPTION
The
.Sy zstream
utility manipulates ZFS send streams output by the
.Sy zfs send
command.
.Bl -tag -width ""
.It Xo
.Nm
.Cm dump
.Op Fl Cvd
.Op Ar file
.Xc
Print information about the specified send stream, including headers and
record counts.
The send stream may either be in the specified
.Ar file ,
or provided on standard input.
.Bl -tag -width "-D"
.It Fl C
Suppress the validation of checksums.
.It Fl v
Verbose.
Print metadata for each record.
.It Fl d
Dump data contained in each record.
Implies verbose.
.El
.Pp
The
.Nm zstreamdump
alias is provided for compatibility and is equivalent to running
.Nm
.Cm dump .
.It Xo
.Nm
.Cm token
.Ar resume_token
.Xc
Dumps zfs resume token information
.It Xo
.Nm
.Cm decompress
.Op Fl v
.Op Ar object Ns Sy \&, Ns Ar offset Ns Op Sy \&, Ns Ar type Ns ...
.Xc
Decompress selected records in a ZFS send stream provided on standard input,
when the compression type recorded in ZFS metadata may be incorrect.
Specify the object number and byte offset of each record that you wish to
decompress.
Optionally specify the compression type.
Valid compression types include
.Sy off ,
.Sy gzip ,
.Sy lz4 ,
.Sy lzjb ,
.Sy zstd ,
and
.Sy zle .
The default is
.Sy lz4 .
Every record for that object beginning at that offset will be decompressed, if
possible.
It may not be possible, because the record may be corrupted in some but not
all of the stream's snapshots.
Specifying a compression type of
.Sy off
will change the stream's metadata accordingly, without attempting decompression.
This can be useful if the record is already uncompressed but the metadata
insists otherwise.
The repaired stream will be written to standard output.
.Bl -tag -width "-v"
.It Fl v
Verbose.
Print summary of decompressed records.
.El
.It Xo
.Nm
.Cm drop_record
.Op Fl v
.Op Ar object Ns Sy \&, Ns Ar offset Ns Op Sy \& ...
.Xc
Drop selected records from a ZFS send stream provided on standard input,
specified by object number and byte offset.
Only WRITE and WRITE_EMBEDDED are records are supported, currently.
The repaired stream will be written to standard output.
.Bl -tag -width "-v"
.It Fl v
Verbose.
Print summary of dropped records.
.El
.It Xo
.Nm
.Cm redup
.Op Fl v
.Ar file
.Xc
Deduplicated send streams can be generated by using the
.Nm zfs Cm send Fl D
command.
The ability to send deduplicated send streams is deprecated.
In the future, the ability to receive a deduplicated send stream with
.Nm zfs Cm receive
will be removed.
However, deduplicated send streams can still be received by utilizing
.Nm zstream Cm redup .
.Pp
The
.Nm zstream Cm redup
command is provided a
.Ar file
containing a deduplicated send stream, and outputs an equivalent
non-deduplicated send stream on standard output.
Therefore, a deduplicated send stream can be received by running:
.Dl # Nm zstream Cm redup Pa DEDUP_STREAM_FILE | Nm zfs Cm receive No …
.Bl -tag -width "-D"
.It Fl v
Verbose.
Print summary of converted records.
.El
.It Xo
.Nm
.Cm raw
.Op Fl v
.Op Fl g Ar fromguid
.Ar image|device
.Xc
Apply a zvol send stream to a raw image or device.
When
.Ar fromguid
is provided, the initial fromguid of the stream will be checked to ensure it
matches the given value.
The final toguid of the stream is printed to stdout on completion.
With
.Fl v ,
details of the records in the stream are printed in similar fashion to
.Nm
.Cm dump .
.It Xo
.Nm
.Cm recompress
.Op Fl t Ar num_threads
.Op Fl l Ar level
.Ar algorithm
.Xc
Recompresses a send stream from standard input using the specified compression
.Ar algorithm
and optional
.Ar level ,
and writes the modified stream to standard output.
All WRITE records in the send stream will be compressed unless compression
fails to reduce their size compared to being left uncompressed.
The
.Ar algorithm
can be any valid value to the
.Nm compress
property.
Note that encrypted send streams cannot be recompressed.
.Bl -tag -width "-t"
.It Fl t Ar num_threads
Specifies the number of compression worker threads.
By default,
.Nm zstream Cm recompress
creates a thread for every CPU core.
It's generally unproductive to raise
.Ar num_threads
above this default.
On some architectures, contention for resources such as memory caches can
reduce efficiency when many threads are active.
On those systems, it may be helpful to run with fewer threads,
particularly when automating operations on shared servers.
.El
.Bl -tag -width "-l"
.It Fl l Ar level
Specifies compression level.
Only needed for algorithms where the level is not implied as part of the name
of the algorithm (e.g. gzip-3 does not require it, while zstd does, if a
non-default level is desired).
.El
.El
.
.Sh EXAMPLES
.Ss Recovering from OpenZFS bug #12762
First, determine which records are corrupt.
That cannot be done automatically; it requires information beyond ZFS's
metadata.
If object
.Sy 128
is corrupted at offset
.Sy 0
and is compressed using
.Sy lz4 ,
then run this command:
.Bd -literal
.No # Nm zfs Ar send Fl c Ar … | Nm zstream decompress Ar 128,0,lz4 | \
Nm zfs recv Ar …
.Ed
.
.Ss Recovering from OpenZFS bug #18239
The bogus records typically have an absurdly large offset, and can be seen with
a command like
.Nm zdb Fl ddddd Ar dataset Ar object
or
.Nm zstream Ar dump Fl v .
To recover, send the dataset and use
.Nm zstream
to drop the bogus record, then receive into a new dataset.
.Bd -literal
.No # Nm zfs Ar send Ar ... | Nm zstream drop_record Ar 3545761,18446744073709486080 | \
Nm zfs recv Ar ...
.Ed
.
.Ss Migrating from a zvol to a physical disk
Given a zvol tank/vol with snapshots @1 @2 and @3, and an adequately sized
physical disk /dev/da4:
.Bd -literal
.No # Nm zfs Cm send Fl Lec Ar tank | Nm zstream Cm raw Ar /dev/da4
16731506615198184313
.No # Nm zfs Cm send Fl Lec Fl I Ar @1 Ar tank/vol@3 | Nm zstream Cm raw Fl g Ar 16731506615198184313 Ar /dev/da4
10690368765373298656
.Ed
.Sh SEE ALSO
.Xr zdb 8 ,
.Xr zfs 8 ,
.Xr zfs-receive 8 ,
.Xr zfs-send 8 ,
.Lk https://github.com/openzfs/zfs/issues/12762
.Lk https://github.com/openzfs/zfs/issues/18239


cmd/zstream/zstream.c
// SPDX-License-Identifier: CDDL-1.0
/*
 * CDDL HEADER START
 *
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 *
 * CDDL HEADER END
 */

/*
 * Copyright (c) 2020 by Delphix. All rights reserved.
 * Copyright (c) 2020 by Datto Inc. All rights reserved.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "zstream.h"

void
zstream_usage(void)
{
	(void) fprintf(stderr,
	    "usage: zstream command args ...\n"
	    "Available commands are:\n"
	    "\n"
	    "\tzstream dump [-vCd] FILE\n"
	    "\t... | zstream dump [-vCd]\n"
	    "\n"
	    "\tzstream decompress [-v] [OBJECT,OFFSET[,TYPE]] ...\n"
	    "\n"
	    "\tzstream drop_record [-v] [OBJECT,OFFSET] ...\n"
	    "\n"
	    "\tzstream raw [-v] [-g guid] FILE|DEV FILE\n"
	    "\t... | zstream raw [-v] [-g guid] FILE|DEV\n"
	    "\n"
	    "\tzstream recompress [-t num_threads] [-l level] TYPE\n"
	    "\n"
	    "\tzstream token resume_token\n"
	    "\n"
	    "\tzstream redup [-v] FILE | ...\n");
	exit(1);
}

int
main(int argc, char *argv[])
{
	char *basename = strrchr(argv[0], '/');
	basename = basename ? (basename + 1) : argv[0];
	if (argc >= 1 && strcmp(basename, "zstreamdump") == 0)
		return (zstream_do_dump(argc, argv));

	if (argc < 2)
		zstream_usage();

	char *subcommand = argv[1];

	if (strcmp(subcommand, "dump") == 0) {
		return (zstream_do_dump(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "decompress") == 0) {
		return (zstream_do_decompress(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "drop_record") == 0) {
		return (zstream_do_drop_record(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "raw") == 0) {
		return (zstream_do_raw(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "recompress") == 0) {
		return (zstream_do_recompress(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "token") == 0) {
		return (zstream_do_token(argc - 1, argv + 1));
	} else if (strcmp(subcommand, "redup") == 0) {
		return (zstream_do_redup(argc - 1, argv + 1));
	} else {
		zstream_usage();
	}
}
